File: 0039.CombinationSum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        ans = []
        self.dfs(sorted(candidates), target, 0, [], ans)
        return ans

    def dfs(self, cand, target, index, temp, ans):
        if target == 0:
            ans.append(temp)
            return
        for i in xrange(index, len(cand)):
            if target < cand[i]:
                return
            self.dfs(cand, target-cand[i], i, temp+[cand[i]], ans)

File: 0000-0999/0039.CombinationSum.go
func combinationSum(candidates []int, target int) [][]int {
    ans := [][]int{}
    dfs(candidates, []int{}, target, 0, &ans)
    return ans
}

func dfs(cand, path []int, target, idx int, ans *[][]int) {
    if target < 0 {
        return
    }
    if target == 0 {
        tmp := make([]int, len(path))
        copy(tmp, path)
        *ans = append(*ans, tmp)
        return
    }
    for i := idx; i < len(cand); i++ {
        dfs(cand, append(path, cand[i]), target-cand[i], i, ans)
    }
}
